File: app/src/main/java/com/scottmangiapane/cardclicks/GameActivity.java
package com.scottmangiapane.cardclicks;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.activity.EdgeToEdge;
import androidx.activity.OnBackPressedCallback;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.preference.PreferenceManager;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import java.util.Random;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

public class GameActivity extends AppCompatActivity {
    private static final int[] CARD_IDS = {
            R.id.card_01, R.id.card_02, R.id.card_03, R.id.card_04,
            R.id.card_05, R.id.card_06, R.id.card_07, R.id.card_08,
            R.id.card_09, R.id.card_10, R.id.card_11, R.id.card_12};
    private static final long GAME_DURATION = 121000;

    private final Random generator = new Random();

    private List<ImageView> cards;
    private List<ImageView> selectedCards;
    private CardViewModel vm;

    private CountDownTimer timer;
    private SharedPreferences sp;

    private TextView buttonPauseRestart;
    private TextView buttonStopExit;
    private TextView scoreView;
    private TextView timeCount;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_game);

        cards = Arrays
                .stream(CARD_IDS)
                .mapToObj(id -> (ImageView) findViewById(id))
                .collect(Collectors.toList());
        cards.forEach(b -> b.setOnClickListener(v -> onButtonClicked((ImageView) v)));
        selectedCards = new ArrayList<>();
        vm = new ViewModelProvider(this, new CardViewModelFactory(CARD_IDS.length, GAME_DURATION))
                .get(CardViewModel.class);

        sp = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());

        buttonPauseRestart = findViewById(R.id.button_pause_restart);
        buttonPauseRestart.setOnClickListener(v -> pauseGame());
        buttonStopExit = findViewById(R.id.button_stop_exit);
        buttonStopExit.setOnClickListener(v -> endGame());
        scoreView = findViewById(R.id.final_score);
        timeCount = findViewById(R.id.time_count);

        refreshViews();
        if (!vm.isGameRunning)
            endGame();

        getOnBackPressedDispatcher().addCallback(this, new OnBackPressedCallback(true) {
            @Override
            public void handleOnBackPressed() {
                if (vm.isGameRunning)
                    pauseGame();
            }
        });
    }

    @Override
    public void onPause() {
        super.onPause();
        if (timer != null)
            timer.cancel();
    }

    @Override
    public void onResume() {
        super.onResume();
        if (vm.isGameRunning) {
            timer = new CountDownTimer(vm.timeRemaining, 1000) {
                public void onTick(long milliseconds) {
                    vm.timeRemaining = milliseconds;
                    timeCount.setText(String.format(
                            Locale.getDefault(),
                            "%d:%02d",
                            TimeUnit.MILLISECONDS.toMinutes(milliseconds),
                            TimeUnit.MILLISECONDS.toSeconds(milliseconds) % 60));
                }

                public void onFinish() {
                    endGame();
                }
            }.start();
        }
    }

    @Override
    public void recreate () {
        super.recreate();
        vm.reset();
    }

    private void endGame() {
        vm.isGameRunning = false;
        if (timer != null)
            timer.cancel();

        buttonPauseRestart.setBackgroundColor(getColor(R.color.colorDanger));
        buttonPauseRestart.setText(getString(R.string.restart));
        buttonPauseRestart.setOnClickListener(v -> this.recreate());
        buttonStopExit.setText(getString(R.string.exit));
        buttonStopExit.setOnClickListener(v -> {
            finish();
            overridePendingTransition(R.transition.slide_right_1, R.transition.slide_right_2);
        });
        selectedCards.clear();

        for (ImageView button : cards)
            button.setAlpha(.5f);
        List<Integer[]> sets = vm.game.getSets();
        Integer[] set = sets.get(generator.nextInt(sets.size()));
        for (int index : set)
            cards.get(index).setAlpha(1f);

        scoreView.setText(String.format(
                Locale.getDefault(),
                "%s · %s",
                getString(R.string.score, vm.score),
                getString(R.string.high, sp.getInt("score_1", 0))));
        timeCount.setText(R.string.empty_timer);
    }

    private void onButtonClicked(ImageView button) {
        if (!vm.isGameRunning) return;

        if (selectedCards.contains(button)) {
            selectedCards.remove(button);
            refreshViews();
            return;
        }
        selectedCards.add(button);
        refreshViews();

        if (selectedCards.size() == 3) {
            if (vm.game.isSet(
                    cards.indexOf(selectedCards.get(0)),
                    cards.indexOf(selectedCards.get(1)),
                    cards.indexOf(selectedCards.get(2)))) {
                String scoreText = getString(R.string.score, ++vm.score);
                scoreView.setText(scoreText);
                saveScore();
                vm.game.rotateCards(
                        cards.indexOf(selectedCards.get(0)),
                        cards.indexOf(selectedCards.get(1)),
                        cards.indexOf(selectedCards.get(2)));
            }
            selectedCards.clear();
            refreshViews();
        }
    }

    private void pauseGame() {
        Intent intent = new Intent(GameActivity.this, PausedActivity.class);
        startActivity(intent);
    }

    private void refreshViews() {
        for (int i = 0; i < cards.size(); i++) {
            cards.get(i).setAlpha((selectedCards.isEmpty()) ? 1f : .5f);
            cards.get(i).setImageResource(vm.game.get(i).drawable);
        }
        for (ImageView button : selectedCards)
            button.setAlpha(1f);
        scoreView.setText(getString(R.string.score, vm.score));
    }

    private void saveScore() {
        SharedPreferences.Editor editor = sp.edit();
        List<Integer> scores = new ArrayList<>();
        scores.add(sp.getInt("score_1", 0));
        scores.add(sp.getInt("score_2", 0));
        scores.add(sp.getInt("score_3", 0));
        scores.add(vm.score);
        scores.sort((a, b) -> b - a);
        editor.putInt("score_1", scores.get(0));
        editor.putInt("score_2", scores.get(1));
        editor.putInt("score_3", scores.get(2));
        editor.apply();
    }
}

File: app/src/main/java/com/scottmangiapane/cardclicks/GameActivity.kt
package com.scottmangiapane.cardclicks

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.os.CountDownTimer
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.preference.PreferenceManager
import java.util.Arrays
import java.util.Locale
import java.util.concurrent.TimeUnit
import java.util.function.Consumer
import java.util.stream.Collectors
import androidx.core.content.edit

class GameActivity : AppCompatActivity() {
    private lateinit var cards: List<ImageView>
    private val selectedCards: MutableList<ImageView> = ArrayList()
    private lateinit var vm: CardViewModel
    private var gameTimer: CountDownTimer? = null
    private var showSetsTimer: CountDownTimer? = null
    private val sp: SharedPreferences by lazy { PreferenceManager.getDefaultSharedPreferences(applicationContext) }
    private lateinit var buttonPauseRestart: TextView
    private lateinit var buttonStopExit: TextView
    private lateinit var scoreView: TextView
    private lateinit var timeCount: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_game)
        cards = Arrays
            .stream(CARD_IDS)
            .mapToObj { id: Int -> findViewById<View>(id) as ImageView }
            .collect(Collectors.toList())
        cards.forEach(Consumer { b: ImageView -> b.setOnClickListener { v: View -> onButtonClicked(v as ImageView) } })
        vm = ViewModelProvider(this, CardViewModelFactory(CARD_IDS.size, GAME_DURATION))[CardViewModel::class.java]
        buttonPauseRestart = findViewById(R.id.button_pause_restart)
        buttonPauseRestart.setOnClickListener { pauseGame() }
        buttonStopExit = findViewById(R.id.button_stop_exit)
        buttonStopExit.setOnClickListener { endGame() }
        scoreView = findViewById(R.id.final_score)
        timeCount = findViewById(R.id.time_count)
        refreshViews()
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                if (vm.isGameRunning) pauseGame()
            }
        })
    }

    public override fun onPause() {
        super.onPause()
        gameTimer?.cancel()
        showSetsTimer?.cancel()
    }

    public override fun onResume() {
        super.onResume()
        if (!vm.isGameRunning) {
            endGame()
            return
        }
        gameTimer = object : CountDownTimer(vm.timeRemaining, 200) {
            override fun onTick(milliseconds: Long) {
                vm.timeRemaining = milliseconds
                timeCount.text = String.format(
                    Locale.getDefault(),
                    "%d:%02d",
                    TimeUnit.MILLISECONDS.toMinutes(milliseconds),
                    TimeUnit.MILLISECONDS.toSeconds(milliseconds) % 60
                )
            }

            override fun onFinish() {
                endGame()
            }
        }.start()
    }

    override fun recreate() {
        super.recreate()
        vm.reset()
    }

    private fun endGame() {
        vm.isGameRunning = false
        gameTimer?.cancel()
        buttonPauseRestart.setBackgroundColor(getColor(R.color.colorDanger))
        buttonPauseRestart.text = getString(R.string.restart)
        buttonPauseRestart.setOnClickListener { recreate() }
        buttonStopExit.text = getString(R.string.exit)
        buttonStopExit.setOnClickListener { finish() }
        selectedCards.clear()
        scoreView.text = String.format(
            Locale.getDefault(),
            "%s · %s",
            getString(R.string.score, vm.score),
            getString(R.string.high, sp.getInt("score_1", 0))
        )
        timeCount.setText(R.string.empty_timer)
        showSets()
    }

    private fun onButtonClicked(button: ImageView) {
        if (!vm.isGameRunning) return
        if (selectedCards.contains(button)) {
            selectedCards.remove(button)
            refreshViews()
            return
        }
        selectedCards.add(button)
        refreshViews()
        if (selectedCards.size == 3) {
            if (vm.game.isSet(
                    cards.indexOf(selectedCards[0]),
                    cards.indexOf(selectedCards[1]),
                    cards.indexOf(selectedCards[2])
                )
            ) {
                val scoreText = getString(R.string.score, ++vm.score)
                scoreView.text = scoreText
                saveScore()
                vm.game.rotateCards(
                    cards.indexOf(selectedCards[0]),
                    cards.indexOf(selectedCards[1]),
                    cards.indexOf(selectedCards[2])
                )
            }
            selectedCards.clear()
            refreshViews()
        }
    }

    private fun pauseGame() {
        val intent = Intent(this@GameActivity, PausedActivity::class.java)
        startActivity(intent)
    }

    private fun refreshViews() {
        for (i in cards.indices) {
            cards[i].alpha = if (selectedCards.isEmpty()) 1f else .5f
            cards[i].setImageResource(vm.game[i].drawable)
        }
        for (button in selectedCards) button.alpha = 1f
        scoreView.text = getString(R.string.score, vm.score)
    }

    private fun saveScore() {
        sp.edit {
            val scores: MutableList<Int> = ArrayList()
            scores.add(sp.getInt("score_1", 0))
            scores.add(sp.getInt("score_2", 0))
            scores.add(sp.getInt("score_3", 0))
            scores.add(vm.score)
            scores.sortWith { a: Int, b: Int -> b - a }
            putInt("score_1", scores[0])
            putInt("score_2", scores[1])
            putInt("score_3", scores[2])
        }
    }

    private fun showSets() {
        showSetsTimer = object : CountDownTimer(1000, 1000) {
            private val sets = vm.game.getSets()
            private var index = 0
            override fun onTick(milliseconds: Long) {
                for (button in cards) button.alpha = .5f
                val set = sets[index++ % sets.size]
                for (index in set) cards[index].alpha = 1f
            }

            override fun onFinish() {
                showSetsTimer?.start()
            }
        }.start()
    }

    companion object {
        private val CARD_IDS = intArrayOf(
            R.id.card_01, R.id.card_02, R.id.card_03, R.id.card_04,
            R.id.card_05, R.id.card_06, R.id.card_07, R.id.card_08,
            R.id.card_09, R.id.card_10, R.id.card_11, R.id.card_12
        )
        private const val GAME_DURATION: Long = 121000
    }
}
